File: 78-subsets/subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        
        vector<vector<int>> ans;
        vector<int> subset;

        helper(ans,nums,0,subset);
        return ans;
    }

private:

    void helper(vector<vector<int>>& ans,vector<int>& nums,int index,vector<int>& subset)
    {
        if(index >= nums.size())
        {
            ans.push_back(subset);
            return;
        }

        subset.push_back(nums[index]);
        helper(ans,nums,index + 1,subset);

        subset.pop_back();
        helper(ans,nums,index + 1,subset);
    }


};

File: 78-subsets/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        n = len(nums)

        def helper(temp,i):
            if i == n:
                ans.append(temp.copy())
                return
            
            temp.append(nums[i])
            helper(temp,i + 1)

            temp.pop()
            helper(temp,i + 1)

    
        helper([],0)
        return ans
